syntax = "proto3";

package Envoy.Extensions.Config;

import "google/protobuf/timestamp.proto";
import "envoy/service/discovery/v3/discovery.proto";

// [#not-implemented-hide:]
// Serialized xDS configuration fetched and persisted by the Envoy instance from one or more
// xDS management servers.
// [#next-free-field: 3]
message SavedXdsConfig {
  message XdsServerSavedConfig {
    message ResourceList {
      repeated envoy.service.discovery.v3.Resource resources = 1;
    };

    // Maps the type URL to the resources delivered for that type.
    map<string, ResourceList> per_type_resources = 1; 
  };

  // Maps control plane (i.e. xDS management server) ID to xDS config from that server.
  map<string, XdsServerSavedConfig> per_control_plane_config = 1;
  
  // The timestamp at which the xDS configuration was last persisted to local storage.
  google.protobuf.Timestamp last_updated = 2;
};
